### PR TITLE
Hw1

### PR DESCRIPTION
**Pull request by happyc0der**
# Pull Request Description

## Overview
[Provide a brief overview of what this PR accomplishes]

## Type of Change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvements
- [ ] Test updates

## Changes Made
[Describe the changes you've made in detail]

### Added
- [List new features/functionality added]

### Changed
- [List modifications to existing functionality]

### Removed
- [List removed features/functionality]

## Testing
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] E2E tests added/updated
- [ ] All tests passing locally
- [ ] Coverage requirements met (90%+)

## Additional Information
- [ ] mypy static type checking passes
- [ ] ruff formatting and linting passes
- [ ] Documentation has been updated
- [ ] Component specification updated (if needed)

## Screenshots (if applicable)
[Add screenshots here if UI changes were made]

## Related Issues
- Closes #[issue number]
- References #[issue number]

## Deployment Notes
[Any special considerations for deploying this change]

## Reviewer Checklist
- [ ] Code follows project style guidelines
- [ ] Code is well-documented
- [ ] Changes are covered by tests
- [ ] Type hints are properly used
- [ ] Error handling is appropriate
- [ ] Performance implications considered
- [ ] Security implications considered

## Self Review Checklist
Before requesting review, I have:
- [ ] Performed self-review of my code
- [ ] Commented on hard-to-understand areas
- [ ] Updated relevant documentation
- [ ] Verified all tests pass
- [ ] Verified code formatting and linting pass
- [ ] Verified type checking passes

### CODE REVIEW

File: tests/e2e/test_e2e.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+                captured_output,
+                "Logging output is incorrect",
+            )
+            self.assertIn(
+                "ALERT: Value 20 exceeded threshold 10",
+                captured_output,
+                "Notifier output is incorrect",
+            )
+
+    def test_division_flow(self) -> None:
```
Comment (Mirap9615) [inline tests/e2e/test_e2e.py]: Edge case testing such as testing dividing by zero could be valuable here to ensure Calculator correctly raises and logs errors during end-to-end execution. This would also better demonstrate how errors propagate across components.

File: tests/e2e/test_e2e.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
+import io
+import unittest
+from unittest.mock import patch
+
+from src.calculator.calculator import Calculator
+from src.logger.logger import Logger
+from src.notifier.notifier import Notifier
+
+
+class TestEndToEndFlow(unittest.TestCase):
```
Comment (Mirap9615) [inline tests/e2e/test_e2e.py]: There is a lot of repetition in capturing output and checking logs/alerts. Consider using a helper method to reduce duplication. Also the quality of many of these comments can be called into question as they seem to mark single-liners and other relatively trivial processes

**Review by Mirap9615 — COMMENTED** [review]
I like how the project is organized and the files (especially the testing ones such as test_e2e) are mostly cleanly formatted and thorough. There is room for improvement such as covering edge cases better and having higher quality comments, but good job overall!

**Comment by Yize-Liu-NYU** [comment]
### **Overall Feedback for Team 9 from Team 14**
**Strengths**

- Very detailed documentations
- Correctly setup testing structure
- License is present
- Very high quality code
- Detailed pull request template
- Well structured folders
- Inclusion of pre-commit hooks
- Includes documentation for test results
- Incorporates lint checks in pyproject.toml with no separate ruff.toml file
- Successfully incorporated nose2 and uv
- Readme includes documentation for uv, CircleCI, and nose2
- Properly defines component architecture

**Areas for Improvement**

- Improve documentation for the python programs
- Pull request could mention Homework 1 specific information, that is, setting up repository and such, instead of generic template
- Add more testing for edge cases, mentioned by @Mirap9615 
- Improving testing logic by putting frequently repeated code into helper functions, mentioned by @Mirap9615 
- Have all tests use a single pyproject.toml instead of separate ones
- Can add potential use cases to the readme
- Add a clearer label to the src folder

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            uv venv .venv
+            echo 'export PATH=".venv/bin:$PATH"' >> $BASH_ENV
+            echo "export PYTHONPATH=$PWD/src" >> $BASH_ENV
+            source .venv/bin/activate
+
+      - run:
+          name: Install Dependencies
+          command: |
+            uv pip install ".[dev]"
+            uv pip install mypy ruff nose2 coverage
```
Comment (jaydaftari) [inline .circleci/config.yml]: Why are you using uv pip install? Install it without pip

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+      - run:
+          name: Cleanup old coverage data
+          command: |
+            coverage erase
+            rm -f .coverage*
```
Comment (jaydaftari) [inline .circleci/config.yml]: So, You don't need to remove old coverage, every time you push to circle ci, new coverage is created in artifactory related to that pipeline id

File: component.md
Comment (jaydaftari) [inline component.md:1]: Move it inside, .github

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+]
+
+[tool.ruff.format]
+quote-style = "double"
+indent-style = "space"
+line-ending = "lf"
+
+[tool.ruff.lint]
+select = ["ALL"]
+ignore = ["D100", "D101", "D102", "D103", "D104", "D107","D203", "T201", "S101", "PLR2004", "TRY003", "EM101",
```
Comment (jaydaftari) [inline pyproject.toml]: Give one line comment for everything you are ignoring..

File: src/calculator/__init__.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+__all__ = [
+    "Calculator",
+    "add",
+    "default_calculator",
+    "divide",
+    "multiply",
+    "subtract",
+]
+
+# Package metadata
```
Comment (jaydaftari) [inline src/calculator/__init__.py:23]: No need to mention author names everywhere

File: src/calculator/pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+testpaths = ["tests"]
+python_files = ["test_*.py"]
+python_classes = ["Test*"]
+python_functions = ["test_*"]
+
+[tool.coverage.run]
+branch = true
+source = ["src"]
+omit = [
+    "tests/*", ".venv/*", "build/*", "dist/*", "*/test/*"
```
Comment (jaydaftari) [inline src/calculator/pyproject.toml]: Give one line comment why are you omiting

File: src/logger/pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    "ruff>=0.9.6",
+    "pre-commit>=4.1.0"
+]
+
+dependencies = [
+    "coverage>=7.6.12",
+    "mypy>=1.15.0",
+    "nose2>=0.15.1",
+    "ruff>=0.9.6",
+    "pre-commit>=4.1.0"
```
Comment (jaydaftari) [inline src/logger/pyproject.toml]: I don't think pre-commit dependency is required in all of the packages.. Just using this dependency in root .toml is fine

File: src/notifier/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,97 @@
+[project]
```
Comment (jaydaftari) [inline src/notifier/pyproject.toml:1]: All of your package pyproject.toml have same dependency , it is already in root .toml, so not required everywhere, it is just making the code redundant

File: src/notifier/test/test_notifier.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import io
+import unittest
```
Comment (jaydaftari) [inline src/notifier/test/test_notifier.py]: Why are we using unittest, it is extra, just use pytest.

File: src/notifier/test/test_notifier.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+from unittest.mock import patch
+
+from src.notifier import Notifier, __authors__, __version__
```
Comment (jaydaftari) [inline src/notifier/test/test_notifier.py]: Why are you checking if author name is present in main file or not? This should not be in unit test case..

File: test-results/calculator_unit.xml
Comment (jaydaftari) [inline test-results/calculator_unit.xml:1]: Don't push this file here in repo, it should be in artifactory, and you can generate it locally

File: tests/e2e/test_e2e.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
+import io
+import unittest
+from unittest.mock import patch
+
+from src.calculator.calculator import Calculator
```
Comment (jaydaftari) [inline tests/e2e/test_e2e.py]: We have to create the package of calculator, so it should be imported with import calculator, if you have implemented it properly, goes for everything.
